lib/galaxy/security/passwords.py

```diff
@@ -67,29 +67,36 @@ def check_password_PBKDF2(guess, hashed):
 _pack_int = Struct('>I').pack
 
 
-def pbkdf2_bin(data, salt, iterations=1000, keylen=24, hashfunc=None):
+def pbkdf2_bin(data, salt, iterations=COST_FACTOR, keylen=KEY_LENGTH, hashfunc=HASH_FUNCTION):
     """Returns a binary digest for the PBKDF2 hash algorithm of `data`
     with the given `salt`.  It iterates `iterations` time and produces a
-    key of `keylen` bytes.  By default SHA-1 is used as hash function,
+    key of `keylen` bytes.  By default SHA-256 is used as hash function,
     a different hashlib `hashfunc` can be provided.
     """
-    hashfunc = hashfunc or hashlib.sha1
-    mac = hmac.new(smart_str(data), None, hashfunc)
-
-    def _pseudorandom(x, mac=mac):
-        h = mac.copy()
-        h.update(x)
-        digest = h.digest()
-        if six.PY2:
+    if hasattr(hashlib, 'pbkdf2_hmac'):
+        # Use hashlib.pbkdf2_hmac, new in python 2.7.8
+        data = smart_str(data)
+        salt = smart_str(salt)
+        return hashlib.pbkdf2_hmac(hashfunc, data, salt, iterations, keylen)
+    else:
+        # Use the old implementation of hashing as a fallback
+        if isinstance(hashfunc, six.string_types):
+            hashfunc = getattr(hashlib, hashfunc, None)
+        hashfunc = hashfunc or hashlib.sha1
```

Comment:
Given that all callers of this function are in this file, I think this can be simplified.

```suggestion
        hashfunc = getattr(hashlib, hashfunc, None)
```